lib/build.ts

```diff
@@ -187,3 +200,125 @@ export async function runBuildTask(
+//
+//      + Some V1 manifests
+//          -> Result: Docker builds, assuming that the platforms will match up.
+//                     Everything works as long as bin_fmt is enabled correctly.
+//                     Otherwise, an execution error will occur.
+//
+//   So what do we do?
+//
+//      + In the case of all images having a V2 manfifest, we can simply pass the plaform flag.
+//
+//      + In the case of having some V1 manifests ...
+//          - If we have balenalib images, we know that they have been updated.  Throw an error
+//            and tell the user to update
+//          - If we do not have balenalib images, don't pass the platform flag.  Let Docker
+//            close it's eyes and hope for the best.  This opens the possibility for an exec
+//            error, but also allows users to continue to use v1 images if they want to / need to.
+//
+async function getRecommendedPlatformhHandling(
+	task: BuildTask,
+	docker: Dockerode,
+): Promise<'allowEnginePlatformHandling' | 'noPlatformHandling' | 'unknown'> {
+	let imageReferences: string[];
+
+	const getHasV1Manfiest = async (repository: string) => {
+		const registry = new RegistryClient(docker.modem);
+		try {
+			const manifest = await registry.getManifest(repository);
+			return (
+				(manifest as DockerImageManifest).Descriptor.mediaType ===
+				MEDIATYPE_MANIFEST_V1
+			);
+		} catch {
+			// eat the exception... yummy!
+		}
+
+		// at this time, treat any errors as true (v1 manifest)
+		return true;
+	};
+
+	if (task.imageName) {
+		imageReferences = [task.imageName + !!task.tag ? `:${task.tag}` : ''];
+	} else {
+		if (!task.dockerfile) {
+			return 'unknown';
+		}
+		const parsedDockerfile = DockerfileParser.parse(task.dockerfile);
+		const dockerInstructions = parsedDockerfile.getInstructions();
+		const fromInstructions = dockerInstructions.filter(
+			(inst) => inst.getKeyword() === 'FROM',
+		);
+		if (fromInstructions.length === 0) {
+			// bail out, no FROM lines found?
+			return 'unknown';
+		}
+		imageReferences = [];
+		fromInstructions.forEach((inst) => {
+			const fromLineArguments = inst.getArguments();
+			if (fromLineArguments.length === 0) {
+				// bail out, no image specified?
+				return;
+			}
+
+			imageReferences.push(fromLineArguments[0].getValue());
+		});
+	}
+
+	let hasBalenaImages = false;
+	let hasAnyV1Manifest = false;
+
+	await Promise.all(
+		imageReferences.map(async (r) => {
+			const thisImageHasV1Manifest = await getHasV1Manfiest(r);
+			hasAnyV1Manifest = thisImageHasV1Manifest || hasAnyV1Manifest;
+			hasBalenaImages = r.startsWith('balenalib/') || hasBalenaImages;
+		}),
+	);
+
+	if (!hasAnyV1Manifest) {
+		return 'allowEnginePlatformHandling';
+	}
+	if (hasBalenaImages) {
+		throw new BuildProcessError(
+			"Using outdated 'balenalib' images that do not support platform targetting.  Please update to the latest `balenalib` images",
+		);
+	}
+	return 'noPlatformHandling';
```

Comment:
So, in the case of mixed "bad manifest" (hasAnyV1Manifest is true) and "good manifest" multiarch images (say, `debian`), where the "bad manifest" images are **not** balenalib images, this function would return `'noPlatformHandling'`. This could result in the incorrect image architecture being selected for the good manifest images (e.g. Intel instead of ARM), and "exec format error" when the app is deployed to the fleet. In this case, I think we should print a warning message to help the user figure things out:
```
Warning: The following images do not support architecture selection:
- Service "myService1", Image "badrepo1/badimage1"
- Service "myService1", Image "badrepo2/badimage2"

While the following images require architecture selection:
- Service "MyService1", Image, "goodrepo1/goodimage1"
- Service "MyService1", Image, "goodrepo2/goodimage2"

To avoid an image build error, the CLI has disabled architecture selection.
However, this means that the architecture of the machine where the Docker
Engine is running will be used by default to select the architecture of the
images that require architecture selection, which may result in "exec format
error" at runtime. To avoid this problem, please update or replace the images
that do not support architecture selection.
```

Reply:
Ok, good idea.  But the downside is that we lose any short-cutting; gotta check each image now.